lucene/core/src/java/org/apache/lucene/document/IndexSortDocValuesRangeQuery.java

```diff
@@ -0,0 +1,279 @@
+    this.field = field;
+    this.lowerValue = lowerValue;
+    this.upperValue = upperValue;
+  }
+
+  @Override
+  public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
+    return new ConstantScoreWeight(this, boost) {
+      @Override
+      public Scorer scorer(LeafReaderContext context) throws IOException {
+        LeafReader reader = context.reader();
+
+        Sort indexSort = reader.getMetaData().getSort();
+        if (indexSort == null || indexSort.getSort().length == 0) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the index must be sorted.");
+        }
+
+        SortField sortField = indexSort.getSort()[0];
+        if (sortField.getField().equals(field) == false) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the index must be sorted " +
+              "on the same field as the query field.");
+        }
+
+        SortedNumericDocValues sortedDocValues = reader.getSortedNumericDocValues(field);
+        NumericDocValues docValues = DocValues.unwrapSingleton(sortedDocValues);
+        if (docValues == null) {
+          throw new IllegalArgumentException("To use IndexSortDocValuesRangeQuery, the query field must have " +
+              "SortedNumericDocValues and have at most one value per document.");
+        }
+
+        DocIdSetIterator disi = getDocIdSetIterator(lowerValue, upperValue,
+            sortField, context, docValues);
+        return new ConstantScoreScorer(this, score(), scoreMode, disi);
+      }
+
+      @Override
+      public boolean isCacheable(LeafReaderContext ctx) {
+        return DocValues.isCacheable(ctx, field);
+      }
+    };
+  }
+
+  /**
+   * Computes the document IDs that lie within the range [lowerValue, upperValue] by
+   * performing binary search on the field's doc values.
+   *
+   * Because doc values only allow forward iteration, we need to reload the field comparator
+   * every time the binary search accesses an earlier element.
+   */
+  private static DocIdSetIterator getDocIdSetIterator(long lowerValue,
+                                                      long upperValue,
+                                                      SortField sortField,
+                                                      LeafReaderContext context,
+                                                      DocIdSetIterator delegate) throws IOException {
+    int maxDoc = context.reader().maxDoc();
+    if (maxDoc <= 0) {
+      return DocIdSetIterator.empty();
+    }
+
+    if (sortField.getReverse()) {
+      long temp = lowerValue;
+      lowerValue = upperValue;
+      upperValue = temp;
+    }
+
+    // Perform a binary search to find the first document with value >= lowerValue.
+    ValueComparator comparator = loadComparator(sortField, lowerValue, context);
+    int low = 0;
+    int high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
```

Comment:
Let's ensure we take care of the known edge cases of binary ranges here. Maybe add some targeted tests?